skin/src/main/java/org/researchstack/skin/DataProviderErrorReceiver.java

```diff
@@ -0,0 +1,128 @@
+package org.researchstack.skin;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.graphics.Color;
+import android.net.Uri;
+import android.support.annotation.NonNull;
+import android.support.annotation.VisibleForTesting;
+import android.support.design.widget.Snackbar;
+import android.support.v4.content.ContextCompat;
+import android.support.v4.content.LocalBroadcastManager;
+import android.view.View;
+import android.widget.TextView;
+
+import com.google.common.base.Strings;
+
+import org.researchstack.backbone.DataProvider;
+import org.researchstack.backbone.onboarding.OnboardingTaskType;
+import org.researchstack.backbone.utils.LogExt;
+import org.researchstack.skin.ui.BaseActivity;
+
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Created by liujoshua on 11/22/2017.
+ */
+public class DataProviderErrorReceiver extends BroadcastReceiver {
+
+    public static final String ACTION_DATA_PROVIDER_ERROR = "org.researchstack.skin.DataProviderError";
+
+    @Override
+    public void onReceive(Context context, Intent intent) {
+        LogExt.i(getClass(), "Received intent with action: " + intent.getAction());
+
+        if (AppPrefs.getInstance().skippedOnboarding()) {
+            // We don't want to bother a user that has skipped sign-up with the signed out
+            // or consent messages. Short-circuiting until we have an approved message to show
+            // a user that has skipped.
+            return;
+        }
+
+        if (!(context instanceof BaseActivity)) {
+            LogExt.w(getClass(), "Received by class that is not a BaseActivity");
+            return;
+        }
+
+        handleInActivity((BaseActivity) context);
+    }
+
+    public static void handleErrors(@NonNull Context context) {
+        if (!AppPrefs.getInstance().getDataProviderErrors().isEmpty()) {
+            LocalBroadcastManager.getInstance(context)
+                    .sendBroadcast(new Intent(ACTION_DATA_PROVIDER_ERROR));
+        }
+    }
+
+    public static void handleError(@NonNull Context context,
+                                   @NonNull @DataProvider.Errors String error) {
+        checkNotNull(context);
+        checkNotNull(error);
+
+        if (!Strings.isNullOrEmpty(error)) {
+            AppPrefs.getInstance().addDataProviderError(error);
```

Comment:
Did you think about sending the error data as intent data?

Reply:
There maybe not be an activity available to handle the broadcast, so we want to persist the errors. 

I didn't see much advantage to reading in errors and sending them in the message versus using the message as a trigger and reading in the errors.

Also, this edge case seems unlikely, but if unhandled errors built up, they should be handled with different priorities, e.g. re-authentication should happen before re-consent. So we want to ensure we're looking at the set of unhandled errors while the order of the broadcast messages/errors can be totally ignored.